lib/web/apiserver.go

```diff
@@ -492,6 +495,20 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 
 		h.Handle("GET", "/robots.txt", httplib.MakeHandler(serveRobotsTxt))
 		h.Handle("GET", "/web/config.js", h.WithUnauthenticatedLimiter(h.getWebConfig))
+
+		hashFile, err := cfg.StaticFS.Open("/apphash")
+		if err != nil {
+			h.log.WithError(err).Error("Failed to open apphash file. Using version only as etag")
```

Comment:
Some things that caught my attention:

1. If the apphash can't be read, the ETag is technically weak. Again, it's not super important, but why not do it the proper way when we have a chance.
2. [The ETag page on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) says the syntax is as follows below. We don't put the etag value in quotes.

```
ETag: W/"<etag_value>"
ETag: "<etag_value>"
```

3. The log message could be improved from the perspective of a user who just sees the error and has no idea what the hash file is or what the ETag will be used for.
4. If we fail to open the apphash file we fall back to the version ETag. That's reasonable. But when we fail to read the file, we bail out completely – wouldn't it make sense to fall back to the version as well?

The patch below addresses all four items.

<details>
<summary>Patch</summary>

Copy then `pbpaste | git apply` to apply

```diff
diff --git a/lib/web/apiserver.go b/lib/web/apiserver.go
index 8627477a6f..ffbbcf697e 100644
--- a/lib/web/apiserver.go
+++ b/lib/web/apiserver.go
@@ -476,7 +476,7 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 	var indexPage *template.Template
 	// we will set our etag based on the teleport version and
 	// the webasset app hash if available
-	etag := teleport.Version
+	etag := fmt.Sprintf("W/%q", teleport.Version)
 	if cfg.StaticFS != nil {
 		index, err := cfg.StaticFS.Open("/index.html")
 		if err != nil {
@@ -496,18 +496,11 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 		h.Handle("GET", "/robots.txt", httplib.MakeHandler(serveRobotsTxt))
 		h.Handle("GET", "/web/config.js", h.WithUnauthenticatedLimiter(h.getWebConfig))
 
-		hashFile, err := cfg.StaticFS.Open("/apphash")
+		etagFromAppHash, err := readEtagFromAppHash(cfg.StaticFS)
 		if err != nil {
-			h.log.WithError(err).Error("Failed to open apphash file. Using version only as etag")
+			h.log.WithError(err).Error("Could not read apphash from embedded webassets. Using version only as ETag for Web UI assets.")
 		} else {
-			defer hashFile.Close()
-			appHash, err := io.ReadAll(hashFile)
-			if err != nil {
-				h.log.WithError(err).Error("Failed to read contents of the hash file")
-				return nil, trace.Wrap(err)
-			}
-
-			etag = fmt.Sprintf("%s-%s", etag, string(appHash))
+			etag = etagFromAppHash
 		}
 	}
 
@@ -4638,3 +4631,21 @@ func serveRobotsTxt(w http.ResponseWriter, r *http.Request, p httprouter.Params)
 	w.Write([]byte(robots))
 	return nil, nil
 }
+
+func readEtagFromAppHash(fs http.FileSystem) (string, error) {
+	hashFile, err := fs.Open("/apphash")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	defer hashFile.Close()
+
+	appHash, err := io.ReadAll(hashFile)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	versionWithHash := fmt.Sprintf("%s-%s", teleport.Version, string(appHash))
+	etag := fmt.Sprintf("%q", versionWithHash)
+
+	return etag, nil
+}

```
</details>